### DISCUSSION TITLE
elgot + project-name issue

### DISCUSSION THREAD

**Discussion by qcfu-bu** [joaotavora/eglot#1190] [General]
After installing the latest version of eglot (1.13), attempting to start eglot in a buffer generates the error message `Symbol's function definition is void: project-name`.

I saw that another user had a similar issue when trying to run CI on eglot. Apparently eglot depends on `project-name` which was introduced in project.el 0.9.8. The confusing issue that I have is that `find-library` shows that my project.el and project.elc are both on 0.9.8. If I manually evaluate project.el then elgot begins working again.

The following snippet is how project/eglot are set up in my config.
```
(use-package project
  :straight t)

(use-package eglot
  :straight t
  :after project)
```

**Comment by subego** [comment]
<img width="1664" alt="image" src="https://user-images.githubusercontent.com/3454566/227451770-035328dc-c137-40c9-8133-97922f1b6540.png">
In Emacs 28.2, there is no project-name. project-name is available in Emacs 29.0.6.

**Reply by joaotavora** [reply]
Yes, we know this.  But newer project is also available from GNU ELPA https://elpa.gnu.org/packages/project.html and that is an acceptable way to upgrade Eglot's dependencies on Emacs 28. 

For some reason this installation isn't succeeding.

**Comment by joaotavora** [comment]
@qcfu-bu the `use-package` snippets you post are not a good way to debug the issue.  I don't have a good idea of what they do, and even less of how "straight" works.  

I only know (more or less) how `package-install` works, and it works by downloading packages from https://elpa.gnu.org into a directory like `~/.emacs.d/elpa`, and then compiling the packages.

It also sets up autoloads, but does _not_ load the package on instlalation.  

This works, and Eglot's CI tests use this officially supported installation technique.

If you somehow manually _load_ `project.el` before installing Eglot, then the upgrade described above won't work. But why would you do that?? And are you even doing that?? We don't know, because this raunchy `use-package` is hiding all of that.

So, we need the macroexpansion of those `use-package` forms.  Please put the cursor after each of those forms and type `M-x pp-macroexpand-last-sexp`, and then attach that result to this discussion.

**Reply by joaotavora** [reply]
Also, just a wild guess, maybe it'll start working if you remove the first `use-package`, the one for project, replace it with a `(project-initialize)` and restart Emacs.  That's assuming you have nothing else in your configuration.

**Reply by qcfu-bu** [reply]
Thanks for the suggestions. I've did some investigating according to your advice.

1. If eglot is installed in an empty init.el using either package-install or use-package, it just works.
2. If eglot is installed using straight.el, then it does not work.

The straight directive for use-package expands as follows.

```
(use-package eglot
  :straight t)
```

```
(progn
  (straight-use-package 'eglot)
  (defvar use-package--warning71
    #'(lambda
        (keyword err)
        (let
            ((msg
              (format "%s/%s: %s" 'eglot keyword
                      (error-message-string err))))
          (display-warning 'use-package msg :error))))
  (condition-case-unless-debug err
      (if
          (not
           (require 'eglot nil t))
          (display-warning 'use-package
                           (format "Cannot load %s" 'eglot)
                           :error))
    (error
     (funcall use-package--warning71 :catch err))))
```

My guess is that straight.el is preventing the new version of project.el from being loaded as eglot is getting invoked.

**Reply by joaotavora** [reply]
Maybe report this to the to the "straight" maintainer then.

**Reply by qcfu-bu** [reply]
Yeah, sounds good to me.
